Add derived parameters module

cdcapano opened the pull request:
This adds a `derived_parameters.py` module to `pycbc/` that collects a bunch of functions to derive a parameter from one or more other parameters. Some of these functions currently live in pnutils; I've moved those functions over to this module, and pointed the relevant function in pnutils to them. One major difference from pnutils is every function in `derived_parameters` only returns a single parameter (whereas several functions in pnutils return two parameters).

I've also included a set of unit tests, `test/test_derived_parameters.py`, that checks for that returned parameters are consistent with physical boundaries, and does a series of round-robin checks -- e.g., does:
```
mass1_from_mchirp_eta(mchirp_from_mass1_mass2(mass1, mass2), eta_from_mass1_mass2(mass1_mass2)) == mass1
```
(This script helped me to find several bugs in my initial commits.) I also checked that the modified functions in pnutils return the same values after the patch as they do before, on a set of 1000 samples.

This is the first step in making `WaveformArray` being able to deal with non-standard parameters being in fields, such as `mchirp` and `eta` rather than `mass1` and `mass2`. The idea is to add `derived_parameters` public namespace to the namespace available to the `eval` in `FieldArray.__getitem__`, so that these functions may be called. Also, it's useful to have these commonly-used functions in one place.

ahnitz commented on pycbc/derived_parameters.py:
```diff
@@ -0,0 +1,462 @@
+# Copyright (C) 2017  Collin Capano
```
Module can be named a bit shorter. How about just 'conversions' or something even shorter. @cmbiwer Ideas?

cmbiwer replied:
I agree with @ahnitz. ``conversions.py`` sounds good to me.

cdcapano replied:
I have no opinion on module name, so whatever you and @cmbiwer like best, I'm fine to change.

ahnitz commented on pycbc/derived_parameters.py:
```diff
@@ -0,0 +1,462 @@
+    return _formatreturn(ms)
+
+
+def mtotal_from_mass1_mass2(mass1, mass2):
```
All these seem to be the opposite convention as pnutils? Any reason here?

cdcapano replied:
Because I usually find myself in the position of wanting a particular parameter and looking for what function I can use to get it, rather than having some set of parameters and wondering what parameters I can derive using them. For example, if you want `mchirp`, you can just do `derived_parameters.mchirp_` + tab complete in ipython to see what functions will let you get that parameter. I never liked the opposite convention used in pnutils for that reason.

ahnitz commented on pycbc/derived_parameters.py:
```diff
@@ -0,0 +1,462 @@
+#
+# =============================================================================
+#
+def _ensurearray(arg):
```
It seems odd to me that you are forcing everything to be an array. That makes it non-generic and if I pass in floats on the command line, I get an array back, weird, and unexpected I think. Is this necessary? If not, I would we do not do this.

cmbiwer replied:
@ahnitz I think Collin has done what you're asking. He's just doing all the internal math as arrays. Look at the ``_formatreturn`` function.

Testing it out, seems like it works:
```
In [2]: derived_parameters.mass2_from_mchirp_q(1.4, 0.25)
Out[2]: 0.8407871037870017
```

cdcapano replied:
It's only used in the `primary_` and `secondary_` functions. In those I need to do truth testing, which is easier to do if the input is guaranteed to be an array. Note that I wrap the output of those functions `_formatreturn` though, which should return a float if a float was passed in. When you say that if you pass in floats on the command line, you get an array back, which function are you referring to?

It occurs to me that I could just code up the `primary_` and `secondary_` functions using standard python boolean comparisons, then vectorize them. Do you prefer that?

ahnitz commented on pycbc/pnutils.py:
```diff
@@ -30,33 +30,36 @@
 import lal, lalsimulation
 import numpy
 from scipy.optimize import bisect, brentq, minimize
+from pycbc import derived_parameters
```
I assume use of derived_parameters within pnutils is to just keep things temporarily working. I'm not super opposed to this, but it's obviously inelegant. If we go this route, please add an issue and someone needs to work to remove this in the near future. The functions should be all removed from pnutils.

cdcapano replied:
Yeah, I didn't want to go searching through pycbc and replace all instances of pnutils that call these functions at the moment. Sure, I'll open an issue after this patch is pushed.

ahnitz commented on test/test_derived_parameters.py:
```diff
@@ -0,0 +1,147 @@
+import numpy
```
Thank you for this! No comment. Just glad people are adding test cases! Kudos!

cmbiwer replied:
I agree its nice. Should it be apart of ``pycbc/tools/run_travis.sh`` though. I ran the test and it only took 0.26 seconds.

ahnitz commented:
OK, I'm convinced by that.

On Feb 9, 2017 7:52 PM, "Collin Capano" <notifications@github.com> wrote:

> *@cdcapano* commented on this pull request.
> ------------------------------
>
> In pycbc/derived_parameters.py
> <https://github.com/ligo-cbc/pycbc/pull/1429>:
>
> > +    return _formatreturn(mp)
> +
> +
> +def secondary_mass(mass1, mass2):
> +    """Returns the smaller of mass1 and mass2 (s = secondary)."""
> +    mass1 = _ensurearray(mass1)
> +    mass2 = _ensurearray(mass2)
> +    if mass1.shape != mass2.shape:
> +        raise ValueError("mass1 and mass2 must have same shape")
> +    ms = copy.copy(mass2)
> +    mask = mass1 < mass2
> +    ms[mask] = mass1[mask]
> +    return _formatreturn(ms)
> +
> +
> +def mtotal_from_mass1_mass2(mass1, mass2):
>
> Because I usually find myself in the position of wanting a particular
> parameter and looking for what function I can use to get it, rather than
> having some set of parameters and wondering what parameters I can derive
> using them. For example, if you want mchirp, you can just do
> derived_parameters.mchirp_ + tab complete in ipython to see what
> functions will let you get that parameter. I never liked the opposite
> convention used in pnutils for that reason.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ligo-cbc/pycbc/pull/1429>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACGrRp_BKJN_PqzLGR_xPsqjYJQaIAJmks5ra2BdgaJpZM4L8c_4>
> .
>

cdcapano commented:
@cmbiwer @ahnitz Ok, I renamed the module to `conversions.py`, and added the unit test to `run_travis.sh`. (Never did that before... hope I did it right.) @ahnitz Let me know if you want any changes about the `_ensurearray`/`_formatreturn` thing.

ahnitz commented:
@cdcapano Nope, I just misunderstood. No additional comments.

cmbiwer commented:
Thanks for addressing those comments Collin. The landscape diff seems to be wonky but all the other checks have passed.

tdent commented:
Hi Collin, I'm not sure what the appropriate way to proceed on copyright is here.  It looks like you have c&p'd a bunch of functions out of pnutils, in addition to other functionality, then I can't see how it's necessarily OK to claim copyright for only yourself as the author of the 'new' module.  I'm pretty sure that many other people contributed to the pnutils functions.

duncan-brown commented:
@cdcapano I have run into an issue with circular imports between pnutils and conversions, which means that I need to move some pnutils to conversions. In the function
```
def ensurearray(arg):
    """Ensures that the given argument is a numpy array. If it is not, the
    argument is converted to an array.
    """
    input_is_array = isinstance(arg, numpy.ndarray)
    if not input_is_array:
        arr = numpy.array(arg)
        arg = arr
    return arg, input_is_array
```

why do you say `arr = numpy.array(arg)` and not `arr = numpy.array([arg])`? There is a difference between the two (shape and len work with the latter but not the former if arg is a float, for example.).

duncan-brown commented:
If this is actually an issue, I'll open a pr.

cdcapano commented:
I do that because if the input is already a list, you'll end up with a two-dimensional array where you expected a 1D array. For example:
```
>>> inp = [10., 11.]
>>> a = numpy.array([inp])
>>> a
array([[ 10.,  11.]])
>>> a.shape
(1, 2)
>>> a = numpy.array(inp)
>>> a.shape
(2,)
```
If the input is a float, it's true that `len(a)` won't work on it, but shape and size both do. For example:
```
>>> inp = 10.
>>> a = numpy.array(inp)
>>> a.shape
()
>>> a.size
1
```
So, the code should be able to deal with such an array. What is the code that is giving the problem? Is it because you're trying to use `len`? If so, switch to using `.shape` or `.size`. This necessitates the inputs being arrays but wrapping all inputs with `ensurearray` will do that for you.

duncan-brown commented:
Shouldn't the code handle this properly by doing `isinstance(x,list)` as in
```
import numpy
x = 3.

if isinstance(x,list):
    x = numpy.array(x)
else:
    x = numpy.array([x])

print type(x), len(x), x.shape, x.size

y = [3.,4.,5.,6.]

if isinstance(y,list):
    y = numpy.array(y)
else:
    y = numpy.array([y])

print type(y), len(y), y.shape, y.size
```

`ensurearray()` is not ensuring that the return value is consistent in all cases.

cdcapano commented:
I guess that could work, although you'd need to make sure that you're still checking that the input is also not an array, so:
```
if not input_is_array:
    if isinstance(y,list):
        y = numpy.array(y)
    else:
        y = numpy.array([y])
    y = numpy.array([y])
```
Still, I'm wary of trying to add all of these if/elses, and of making a change to a function that affects a lot of other things. Can you give an example of the case where `ensurearray()` is not ensureing the return value is consistent?

duncan-brown commented:
Any instance where you want to create an output array inside the conversions function that is the same size as the input arrays. You have to say
```
if not m1.shape:
    m2 = numpy.array(0.)
else:
   m2 = numpy.zeros(m1.size)
```

tdent commented:
I think `if hasattr(arg, '__len__'):` should do what Duncan wants .. 

BTW I also don't see what purpose `arg = arr` serves rather than just `return arr, input_is_array`

duncan-brown commented:
@tdent maybe the code was written by pirates?

cdcapano commented:
No, you can do that...
```
>>> a = numpy.array(10.)
>>> a
array(10.0)
>>> a.shape
()
>>> m2 = numpy.zeros(a.shape)
>>> m2
array(0.0)
```
To set values you can then either do: `m2[()] = 4.`, `m2.itemset(4.)`, or (since it's zero) `m2 += 4.`. See [https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.scalars.html](numpy scalars) docs for more.

cdcapano commented:
@tdent Yeah, not sure why I did that either... I think it was an edit from an earlier block of code that did more inside that if statement. But yeah, that could be simplified to just `arg = numpy.array(arg)` in the if statement.

duncan-brown commented:
OK, that function is still ugly, though...

cdcapano commented:
@duncan-brown How so?

duncan-brown commented:
It's really `ensure_ndarray_or_array_scalar_depending_on_the_input()`. It would be more logical to have it be `ensurendarray()` and always return an `ndarray`. Then you know that e.g. `len()` always works on the output.

cdcapano commented:
But you shouldn't be using `len()` to begin with... stick with the numpy functions and you'll be fine.

duncan-brown commented:
OK, I'll leave it as is, thanks.
